src/python/pants/backend/jvm/tasks/jar_publish.py

```diff
@@ -813,8 +813,7 @@ def check_targets(self, targets):
     def collect_invalid(publish_target, walked_target):
       for derived_target in walked_target.derived_from_chain:
         derived_by_target[derived_target].add(walked_target)
-      if not walked_target.has_sources() or not walked_target.sources_relative_to_buildroot():
-        invalid[publish_target][walked_target].add('No sources.')
+      # Targets with no sources are considered valid - so we continue to publish them.
```

Comment:
In order for this to resolve the user issue, I think that we'll need to ensure that we actually create an empty jar for the no-sources case.

Reply:
Not aware of the full context, but it is legal/not uncommon to have pom only artifacts. E.g. https://repo1.maven.org/maven2/org/apache/httpcomponents/httpcomponents-parent/10/

Reply:
> The util case linked above is not of the style Yi pointed out, but rather a hack with ~aggregator targets to support a legacy layout / jar contents. 

As you said: this is very useful for legacy targets.

When you split a jar from one jar into many jars, it is important for the original jar/pom to not suddenly cease to exist, for a few reasons:

1. If the jar/pom ceases to exist, then you break consumers that are depending on nightly snapshots of a monorepo, without a deprecation cycle.
2. If the consumers do not have a direct dependency on the old jar, but do have a transitive dependency, you can get in a situation where one subgraph depends on the old merged jar, while another subgraph depends on the split jars. Because there is no shared artifact name between these subgraphs, ivy/maven/coursier will not resolve them for you... if you preserve the legacy jar/pom for a deprecation period, you minimize the chance of this.

----

We would be fine with automagically converting something to a `<packaging>pom</packaging>` artifact if it has no sources, but "guaranteeing a jar even if it is empty" seemed like it involved fewer special cases.